Arriba/Arriba.Web/parts/SearchBox.jsx

```diff
@@ -0,0 +1,158 @@
+            e.stopPropagation();
+            e.preventDefault(); // Prevent cursor moving to beginning of line.
+        }
+        if (e.key === "ArrowDown") {
+            this.setState({ sel: (this.state.sel + 1).clamp(0, this.state.suggestions.length - 1) });
+            e.stopPropagation();
+        }
+        if (e.key === "Enter" || this.state.completionCharacters.includes(e.key)) {
+            var item = this.state.suggestions[this.state.sel];
+
+            var separator = (item.category === "Value" && e.key !== " " ? "" : " ");
+            var suffix = (e.key === "Enter" || e.key === "Tab" || e.key === " ") ? "" : e.key;
+            var newQuery = item.replaceAs || (this.state.completed + item.completeAs + separator + suffix);
+            this.setQuery(newQuery);
+            e.preventDefault(); // Suppress focus tabbing.
+        }
+        if (e.key === "Escape") {
+            this.setState({ suggestions: [] });
+        }
+    },
+    handleClickSuggestion: function (item) {
+        var separator = (item.category === "Value" ? "" : " ");
+        this.setQuery(item.replaceAs || this.state.completed + item.completeAs + separator);
+        this.refs.searchBox.focus();
+    },
+    setQuery: function (query) {
+        this.props.onSearchChange(query);
+
+        if (this.suggestions && query.startsWith(this.state.completed)) {
+            const startsWithCI = (a, b) => a.toUpperCase().startsWith(b.toUpperCase());
+            const incomplete = query.slice(this.state.completed.length);
+            const matching = this.suggestions.filter(s => startsWithCI(s.completeAs.trimIf("["), incomplete));
+            if (matching.length) {
+                this.setState({ suggestions: matching, sel: 0 });
+                return;
+            }
+        }
+
+
+        if (this.lastRequest) this.lastRequest.abort();
+        this.lastRequest = jsonQuery(
+            configuration.url + "/suggest?q=" + encodeURIComponent(query),
+            data => {
+                this.suggestions = data.content.suggestions; // A cache which is later filtered and assigned to state.suggestions.
+                this.setState({
+                    suggestions: data.content.suggestions,
+                    sel: 0,
+                    completed: data.content.complete, 
+                    completionCharacters: data.content.completionCharacters.map(c => ({ "\t": "Tab" })[c] || c),
+                })
+            },
+            (xhr, status, err) => console.error(xhr.url, status, err.toString())
+        );
+    },
+    toggleFavorite: function () {
+        if (!this.props.parsedQuery) return;
+        localStorage.updateJson("favorites", favs => ([] || favs).toggle(this.props.parsedQuery));
+    },
+    render: function () {
+        // Generates a SVG histogram to be displayed behind the completion list.
+        // The path goes counter-clockwise starting from the top-right.
+        var svg = this.state.suggestions.length && (() => {
```

Comment:
Can we factor the SVG code into a separate method, so you can see just the summary of what render does for this class?